### DISCUSSION TITLE
Is a single log parallelized?

### DISCUSSION THREAD

**Discussion by schildbach** [SSLMate/certspotter#86] [Q&A] [answered]
I'm running certspotter on a 4 core Intel N100. I've got troubles to keep up to date on a single log:
https://ct.googleapis.com/logs/us1/argon2024/
On some days, it grows a bit faster than I can parse. On other days, it grows slower.

Neither my internet connection nor my CPU is maxed out, but both used to be maxed out while I was still catching up on multiple logs.
This leads me to suspect a single log is only parsed strictly sequential, using only one core. Is that correct?

**Comment by AGWA** [comment]
Correct, a single log is processed sequentially. I would like to add parallel processing (see #34) but It's not trivial because the Merkle Tree has to be updated serially.

Are you using `-start_at_end` and running certspotter continuously?

**Comment by schildbach** [comment]
Thanks for the pointer!

> Are you using -start_at_end

I started at beginning (yes, I'm crazy) and given the fact it's almost done I assume it's going to stay that way for now.

> and running certspotter continuously?

Almost. I've got a difficult powerline connection that sometime breaks.

I think eventually it will catch up. However if the rate of "certificates issues worldwide" continues to increase (exponentially?), I think it will be difficult even for a more powerful core to stay afloat.

**Comment by AGWA** [comment, marked as answer]
certspotter v0.19.0 downloads entries in parallel and should be able to keep up with the growth of logs
